### DISCUSSION TITLE
multi-select, combo-box очень медленно работают при 500+ строк. multi-select - ExpressionChangedAfterItHasBeenCheckedError

### DISCUSSION THREAD

**Discussion by expdts** [taiga-family/taiga-ui#7415] [Q&A] [answered]
[Пример](https://stackblitz.com/edit/angular-tmuv65-gklrq1?file=src%2Fapp%2Fapp.component.html)

`multi-select` и `combo-box` очень медленно работают (открыть список, найти, выбрать и т.п.) и подвисают, если в них содержится более 500 строк. В моем случае 2000+ строк и их количество постоянно растет, от этого не уйти.

Подскажите пожалуйста, можно ли как-то ускорить работу со строками в `multi-select` и `combo-box`?

В `multi-select` периодески  возникает ошибка, в примере на stackblitz она появляется:
```
Error: NG0100: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value for '_focused': 'false'. Current value: 'true'.. Find more at https://angular.io/errors/NG0100
```

**Comment by waterplea** [comment, marked as answer]
1. В DevMode каждый цикл проверки изменений выполняется дважды и в целом он существенно медленнее, на проде должно работать быстрее.
2. Никто не будет открывать и скроллить 500 элементов. Если у тебя их так много, то в `emptyContent` списка выводи что-нибудь вроде "Начните писать для фильтрации результатов" и начинай показывать уже отфильтрованные пункты только после ввода 2-3 символов.
3. Можешь попробовать заюзать виртуальный скролл ([ComboBox](https://taiga-ui.dev/components/combo-box#virtual), [MultiSelect](https://taiga-ui.dev/components/multi-select#virtual))
4. Сейчас активно разрабатывается следующая мажорка 4.0 с существенной переработкой многих контролов, после её выхода можно вернуться к этому вопросу и если всё ещё будет работать недостаточно быстро, то можно будет подумать где и что оптимизировать

**Reply by expdts** [reply]
@waterplea можете подсказать, где можно посмотреть как работает `emptyContent` в данном случае? в документации ничего не нашел об `emptyContent`

**Reply by waterplea** [reply]
Он у `tui-data-list` и `tui-data-list-wrapper`. Просто что выводить, если нет ни одного `tuiOption` внутри:
https://taiga-ui.dev/components/data-list-wrapper/API

**Reply by expdts** [reply]
@waterplea Спасибо. `emptyContent` к сожалению не подходит. Сравнил работу выпадающего списка в `ngx-mat-select-search`, в нем работает без задержек. Буду ожидать 4 версии, надеюсь в ней будет работать быстро.
